### PR TITLE
Make sure everything gets closed in SpillableHashAggregationBuilder

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 1]
```diff
@@ -189,10 +189,10 @@ private boolean shouldMergeWithMemory(long memorySize)
     @Override
     public void close()
     {
-        if (hashAggregationBuilder != null) {
-            hashAggregationBuilder.close();
-        }
         try (Closer closer = Closer.create()) {
+            if (hashAggregationBuilder != null) {
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: Can `hashAggregationBuilder` even be null? When a new instance of `SpillableHashAggregationBuilder` is constructed `hashAggregationBuilder` is guaranteed to be non-null (through `rebuildHashAggregationBuilder()`), no?

Also when we rebuild a new hash agg builder (in `spillToDisk()`), I guess we should close the existing one so that it releases whatever it has acquired.

Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: @nezihyigitbasi fair point. Indeed it is so, but it's not obvious. Since `close()` is quite sensitive operation, I'd prefer to err on the side of defensive programming. What do you think?

Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: if you think it's not obvious let's keep it as is.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 2]
```diff
@@ -278,9 +278,10 @@ private void rebuildHashAggregationBuilder()
         if (hashAggregationBuilder != null) {
             hashCollisions += hashAggregationBuilder.getHashCollisions();
             expectedHashCollisions += hashAggregationBuilder.getExpectedHashCollisions();
+            hashAggregationBuilder.close();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: make it `null` after you close it

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 3]
```diff
@@ -278,9 +278,10 @@ private void rebuildHashAggregationBuilder()
...
         }
 
-        this.hashAggregationBuilder = new InMemoryHashAggregationBuilder(
+        hashAggregationBuilder = new InMemoryHashAggregationBuilder(
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: this seems unrelated

Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: seems so -- method first operates on `hashAggregationBuilder` (and this is what gets closed) and then assigns to `this.hashAggregationBuilder` -- same thing but looks different

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 2 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: why should i need to?

Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: IMO this a good convention when object got invalidated (closed), because:
 - it makes it eligible for GC
 - it make close idempotent (safe for double close)

Nevermind, up to you.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 4]
```diff
@@ -189,10 +189,10 @@ private boolean shouldMergeWithMemory(long memorySize)
     @Override
     public void close()
     {
-        if (hashAggregationBuilder != null) {
-            hashAggregationBuilder.close();
-        }
         try (Closer closer = Closer.create()) {
+            if (hashAggregationBuilder != null) {
+                closer.register(hashAggregationBuilder::close);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: why `hashAggregationBuilder` is not `Optional` as others?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 3 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: I understand it makes this method coherent, but the commit message all the changes are related with closing, while here is plain refactor.

Anyway, it is a minor thing. Up to you. I would separate this change.

Generaly, I think there is a convention to avoid `this.` prefix if possible (reasonable). Intellij marks that already.

**Comment by kokosing** [comment]
Remember to squash the commits, and make sure that travis passes. Ping me when it become ready to be merged.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 2 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: This is not close, but rather `rebuildHashAggregationBuilder()` method and `hashAggregationBuilder` reference is overwritten with new object immediately afterward, so explicit nulling would make sense only for exceptional cases which I choose not to optimize for.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 4 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: Because, in fact, it's not optional. It's always non-null. See however https://github.com/prestodb/presto/pull/8728#discussion_r133591506 for explanation why i retain if-not-null check in `close`.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 2 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: Sounds good

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 4 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: So IMO instead of using IFs to check if it is not null, you should be using cheskState. What do you think?

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: > Since `close()` is quite sensitive operation, I'd prefer to err on the side of defensive programming.

**Comment by findepi** [comment]
squashed, rebased & travis green
